File: TopCoder/src/main/java/com/leetcode/tree/SumRootToLeaf.java
package com.leetcode.tree;

import java.util.List;

//https://leetcode.com/problems/sum-root-to-leaf-numbers/submissions/

public class SumRootToLeaf {
    private int sum = 0;

    public int sumNumbers(TreeNode root) {
        recursive(root, 0);
        return sum;
    }

    private void recursive(TreeNode root, int current) {
        if (root == null) {
            return;
        }

        if (root.left == null && root.right == null) {
            sum +=  (current*10 + root.val);
            return;
        }

        recursive(root.left, current*10 + root.val);
        recursive(root.right, current*10 + root.val);
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        int result = new SumRootToLeaf().sumNumbers(root);
        System.out.println(result);
    }
}

File: TopCoder/python/SumRootToLeaf.py
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

class Solution:
    def sumNumbers(self, root):
        result = []
        self.recursive(root, result, [])
        return sum(result)

    def recursive(self, root, result, current_path):
        if not root:
            return result

        if not root.left and not root.right:
            current_path += [str(root.val)]
            result.append(int("".join(current_path)))
            return result

        self.recursive(root.left, result, current_path + [str(root.val)])
        self.recursive(root.right, result, current_path + [str(root.val)])

root = TreeNode(1)
root.left = TreeNode(2)
root.right = TreeNode(3)

sol = Solution()
print(sol.sumNumbers(root))